<a href="https://colab.research.google.com/github/vamshidhar23/2303A510H7-PYHTON-FOR-DATA-SCIENCE/blob/main/LAB_2_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Analyze student grades using NumPy arrays, including creating a 2D array, calculating statistics, using slicing and boolean indexing, reshaping the array, and commenting on the results.

## Create a 2d numpy array

### Subtask:
Create a 2D NumPy array to store student marks in 3 subjects.


**Reasoning**:
Create a 2D NumPy array with sample student grades and print it.



In [ ]:
import numpy as np

student_grades = np.array([[85, 90, 78],
                           [92, 88, 95],
                           [78, 85, 80],
                           [90, 82, 87],
                           [88, 95, 91]])

print(student_grades)

[[85 90 78]
 [92 88 95]
 [78 85 80]
 [90 82 87]
 [88 95 91]]


## Calculate basic statistics

### Subtask:
Calculate and display the average, minimum, and maximum marks for all students and subjects.


**Reasoning**:
Calculate and print the average, minimum, and maximum marks for all students and subjects using NumPy functions.



In [ ]:
average_mark = np.mean(student_grades)
min_mark = np.min(student_grades)
max_mark = np.max(student_grades)

print(f"Average mark for all students and subjects: {average_mark:.2f}")
print(f"Minimum mark for all students and subjects: {min_mark}")
print(f"Maximum mark for all students and subjects: {max_mark}")

Average mark for all students and subjects: 86.93
Minimum mark for all students and subjects: 78
Maximum mark for all students and subjects: 95


## Analyze student performance

### Subtask:
Use slicing to retrieve and display marks of specific students.


**Reasoning**:
Use slicing to select and display the marks of specific students as instructed.



In [ ]:
# Select and display the marks of the first student
first_student_marks = student_grades[0, :]
print("Marks of the first student:")
print(first_student_marks)

# Select and display the marks of the third and fourth students
specific_students_marks = student_grades[2:4, :]
print("\nMarks of the third and fourth students:")
print(specific_students_marks)

Marks of the first student:
[85 90 78]

Marks of the third and fourth students:
[[78 85 80]
 [90 82 87]]


## Identify high-scoring students

### Subtask:
Use boolean indexing to find and display students who scored above 80 in any subject.


**Reasoning**:
Create a boolean array and use it to select rows where at least one value is greater than 80.



In [ ]:
above_80_boolean = student_grades > 80
students_above_80 = student_grades[np.any(above_80_boolean, axis=1)]
print("Students who scored above 80 in any subject:")
print(students_above_80)

Students who scored above 80 in any subject:
[[85 90 78]
 [92 88 95]
 [78 85 80]
 [90 82 87]
 [88 95 91]]


## Analyze subject-wise performance

### Subtask:
Reshape the array to analyze performance across subjects.


**Reasoning**:
Reshape the array and calculate the average, minimum, and maximum marks for each subject.



In [ ]:
subject_grades = student_grades.reshape(3, 5).T

average_subject_mark = np.mean(subject_grades, axis=0)
min_subject_mark = np.min(subject_grades, axis=0)
max_subject_mark = np.max(subject_grades, axis=0)

print("Average mark for each subject:", average_subject_mark)
print("Minimum mark for each subject:", min_subject_mark)
print("Maximum mark for each subject:", max_subject_mark)

Average mark for each subject: [86.6 85.6 88.6]
Minimum mark for each subject: [78 78 82]
Maximum mark for each subject: [92 95 95]


## Summarize findings

### Subtask:
Add comments to the code explaining the results of the analysis.


**Reasoning**:
The subtask is to add comments to the previously executed code cells to explain the results. I need to go back through the history of executed cells and add comments where calculations or filtering were performed, explaining the code and interpreting the output.



In [ ]:
# Create a 2D NumPy array to store student marks in 3 subjects.
# Each row represents a student, and each column represents a subject.
student_grades = np.array([[85, 90, 78],
                           [92, 88, 95],
                           [78, 85, 80],
                           [90, 82, 87],
                           [88, 95, 91]])

# Print the created array to verify its structure and content.
print(student_grades)

# Calculate the average mark for all students and subjects using np.mean().
average_mark = np.mean(student_grades)
# Calculate the minimum mark across all students and subjects using np.min().
min_mark = np.min(student_grades)
# Calculate the maximum mark across all students and subjects using np.max().
max_mark = np.max(student_grades)

# Print the calculated statistics, formatted to two decimal places for the average.
# The output shows the overall performance: average=86.93, min=78, max=95.
print(f"Average mark for all students and subjects: {average_mark:.2f}")
print(f"Minimum mark for all students and subjects: {min_mark}")
print(f"Maximum mark for all students and subjects: {max_mark}")

# Use slicing to select and display the marks of the first student (index 0).
# The slice [0, :] selects the first row and all columns.
first_student_marks = student_grades[0, :]
print("Marks of the first student:")
print(first_student_marks) # Output: [85 90 78] - These are the marks of the first student in the three subjects.

# Use slicing to select and display the marks of the third and fourth students (indices 2 and 3).
# The slice [2:4, :] selects rows from index 2 up to (but not including) index 4, and all columns.
specific_students_marks = student_grades[2:4, :]
print("\nMarks of the third and fourth students:")
print(specific_students_marks) # Output: [[78 85 80] [90 82 87]] - These are the marks for the third and fourth students.

# Use boolean indexing to create a boolean array indicating where scores are above 80.
above_80_boolean = student_grades > 80
# Use np.any() with axis=1 to check if any score in a row (for a student) is True (above 80).
# This returns a boolean array where True indicates the student has at least one score above 80.
students_above_80 = student_grades[np.any(above_80_boolean, axis=1)]
# Print the rows (students) where at least one subject score was above 80.
# The output shows the students who meet this criterion.
print("Students who scored above 80 in any subject:")
print(students_above_80)

# Reshape the array to analyze performance across subjects.
# .reshape(3, 5) changes the array to 3 rows and 5 columns.
# .T (transpose) then makes it 5 rows and 3 columns, where each row is a subject and each column is a student's score in that subject.
# Note: This reshaping seems incorrect for analyzing performance *across* subjects where each row is a subject.
# A correct reshaping for subject analysis would likely have subjects as rows and students as columns.
# Let's assume the intent was to have subjects as rows and students as columns after reshaping and transposing.
# The original student_grades has students as rows and subjects as columns.
# To get subjects as rows and students as columns, we just need to transpose the original array.
# Let's correct the reshaping and analysis based on subjects as rows.
subject_grades_correct = student_grades.T

# Calculate the average mark for each subject using np.mean() with axis=1 (across columns, which are now students).
average_subject_mark = np.mean(subject_grades_correct, axis=1)
# Calculate the minimum mark for each subject using np.min() with axis=1.
min_subject_mark = np.min(subject_grades_correct, axis=1)
# Calculate the maximum mark for each subject using np.max() with axis=1.
max_subject_mark = np.max(subject_grades_correct, axis=1)

# Print the calculated statistics for each subject.
# The output shows the average, minimum, and maximum scores for each of the three subjects.
print("Average mark for each subject:", average_subject_mark)
print("Minimum mark for each subject:", min_subject_mark)
print("Maximum mark for each subject:", max_subject_mark)

[[85 90 78]
 [92 88 95]
 [78 85 80]
 [90 82 87]
 [88 95 91]]
Average mark for all students and subjects: 86.93
Minimum mark for all students and subjects: 78
Maximum mark for all students and subjects: 95
Marks of the first student:
[85 90 78]

Marks of the third and fourth students:
[[78 85 80]
 [90 82 87]]
Students who scored above 80 in any subject:
[[85 90 78]
 [92 88 95]
 [78 85 80]
 [90 82 87]
 [88 95 91]]
Average mark for each subject: [86.6 88.  86.2]
Minimum mark for each subject: [78 82 78]
Maximum mark for each subject: [92 95 95]


## Summary:

### Data Analysis Key Findings

*   The average mark across all students and subjects is 86.93, with a minimum mark of 78 and a maximum mark of 95.
*   The marks for the first student are 85, 90, and 78.
*   The marks for the third and fourth students are [[78, 85, 80], [90, 82, 87]].
*   All students scored above 80 in at least one subject.
*   The average marks for the three subjects are approximately 86.6, 88.0, and 86.2, respectively.
*   Subject 2 has the highest average mark (88.0).
*   The minimum marks for the three subjects are 78, 82, and 78, respectively.
*   The maximum marks for the three subjects are 92, 95, and 95, respectively.

### Insights or Next Steps

*   Analyze the performance of individual students to identify those who may need additional support or those who are excelling.
*   Investigate the subject with the lowest average score (Subject 3) to understand potential reasons for lower performance.


# Task
Analyze employee salary data using pandas DataFrames, including filtering, sorting, adding a bonus column, calculating total expenses, and saving the results to a CSV file.

## Create dataframe

### Subtask:
Create a pandas DataFrame with employee Name, ID, Department, and Salary data.


**Reasoning**:
Create a pandas DataFrame from a dictionary containing employee data and display it.



In [ ]:
import pandas as pd

# Create a dictionary with employee data
employee_data = {
    'Name': ['Alice', 'Bob', 'Charlie', 'David', 'Eve'],
    'ID': [101, 102, 103, 104, 105],
    'Department': ['HR', 'IT', 'Finance', 'Marketing', 'IT'],
    'Salary': [60000, 75000, 65000, 70000, 80000]
}

# Convert the dictionary into a pandas DataFrame
employee_df = pd.DataFrame(employee_data)

# Display the DataFrame
display(employee_df)

## Filter by salary

### Subtask:
Filter the DataFrame to include only employees with a salary greater than 50,000.


**Reasoning**:
Use boolean indexing to filter the DataFrame to include only employees with a salary greater than 50,000 and display the result.



In [ ]:
# Create a boolean mask where the 'Salary' column is greater than 50000.
salary_filter = employee_df['Salary'] > 50000

# Apply this boolean mask to the employee_df DataFrame to select only the rows where the condition is True.
filtered_employee_df = employee_df[salary_filter]

# Display the filtered DataFrame.
display(filtered_employee_df)

## Sort by salary

### Subtask:
Sort the DataFrame by salary in descending order.


**Reasoning**:
Sort the filtered DataFrame by the 'Salary' column in descending order and display the result.



In [1]:
# Sort the filtered_employee_df DataFrame by the 'Salary' column in descending order.
sorted_employee_df = filtered_employee_df.sort_values(by='Salary', ascending=False)

# Display the sorted_employee_df DataFrame to verify the sorting.
display(sorted_employee_df)

NameError: name 'filtered_employee_df' is not defined

**Reasoning**:
The previous command failed because the `filtered_employee_df` DataFrame was not available in the current kernel session. I need to recreate the `employee_df` and then `filtered_employee_df` from the previous steps before sorting.



In [2]:
import pandas as pd

# Recreate the original DataFrame
employee_data = {
    'Name': ['Alice', 'Bob', 'Charlie', 'David', 'Eve'],
    'ID': [101, 102, 103, 104, 105],
    'Department': ['HR', 'IT', 'Finance', 'Marketing', 'IT'],
    'Salary': [60000, 75000, 65000, 70000, 80000]
}
employee_df = pd.DataFrame(employee_data)

# Recreate the filtered DataFrame
filtered_employee_df = employee_df[employee_df['Salary'] > 50000]

# Sort the filtered_employee_df DataFrame by the 'Salary' column in descending order.
sorted_employee_df = filtered_employee_df.sort_values(by='Salary', ascending=False)

# Display the sorted_employee_df DataFrame to verify the sorting.
display(sorted_employee_df)

Name   ID Department  Salary
4      Eve  105         IT   80000
1      Bob  102         IT   75000
3    David  104  Marketing   70000
2  Charlie  103    Finance   65000
0    Alice  101         HR   60000

## Add bonus column

### Subtask:
Add a new column to the DataFrame for calculating a 10% bonus based on salary.


**Reasoning**:
Add a new column 'Bonus' to the DataFrame and calculate its values based on the 'Salary' column, then display the updated DataFrame.



In [3]:
# Add a new column 'Bonus' to the sorted_employee_df DataFrame.
# Calculate the bonus as 10% of the salary.
sorted_employee_df['Bonus'] = sorted_employee_df['Salary'] * 0.10

# Display the updated DataFrame to verify the addition of the 'Bonus' column.
display(sorted_employee_df)

Name   ID Department  Salary   Bonus
4      Eve  105         IT   80000  8000.0
1      Bob  102         IT   75000  7500.0
3    David  104  Marketing   70000  7000.0
2  Charlie  103    Finance   65000  6500.0
0    Alice  101         HR   60000  6000.0

## Calculate total salary expense

### Subtask:
Calculate the total salary expense, including the bonuses.


**Reasoning**:
Calculate the total salary and bonus from the DataFrame and print the sum.



In [4]:
# Sum the 'Salary' column to get the total salary.
total_salary = sorted_employee_df['Salary'].sum()

# Sum the 'Bonus' column to get the total bonus amount.
total_bonus = sorted_employee_df['Bonus'].sum()

# Add the total salary and total bonus to get the total salary expense.
total_salary_expense = total_salary + total_bonus

# Print the total salary expense.
print(f"Total salary expense (including bonuses): ${total_salary_expense:.2f}")

Total salary expense (including bonuses): $385000.00


## Save to csv

### Subtask:
Save the final DataFrame with the new bonus column to a CSV file.


**Reasoning**:
Save the DataFrame to a CSV file, excluding the index.



In [5]:
# Save the sorted_employee_df DataFrame to a CSV file named 'employee_salary_with_bonus.csv'.
# Set index=False to exclude the DataFrame index from the saved CSV file.
sorted_employee_df.to_csv('employee_salary_with_bonus.csv', index=False)

print("DataFrame saved to 'employee_salary_with_bonus.csv'")

DataFrame saved to 'employee_salary_with_bonus.csv'


## Summary:

### Data Analysis Key Findings

*   The initial DataFrame contained data for 5 employees, including their Name, ID, Department, and Salary.
*   Filtering the DataFrame for salaries greater than \\$50,000 kept all 5 employees in the dataset.
*   The DataFrame was successfully sorted by Salary in descending order, placing the highest earners first.
*   A new column 'Bonus' was added, calculated as 10% of the 'Salary'.
*   The total salary expense, including the calculated bonuses, is \\$385,000.00.
*   The final DataFrame, including the bonus column and sorted by salary, was saved to a CSV file named 'employee\_salary\_with\_bonus.csv' without the DataFrame index.

### Insights or Next Steps

*   The analysis provides a clear overview of employee salaries and the impact of a 10% bonus on total expenses.
*   Further analysis could involve calculating department-specific salary expenses or exploring other bonus calculation methods.
